build/common/bitbake_util.py

```diff
@@ -67,86 +62,84 @@ def __init__(
 
         Optional args:
-        * timeout (int): how long to wait for the next prompt. None means to
-                         wait indefinitely
-        * stdout (bool): flag to print the output to stdout
+        * kwargs for subprocess.run().
 
         Return:
-        * output (str): output of the specified command
+        * subprocess.CompletedProcess object.
 
         """
-        print('Running "{}"...'.format(command))
-        output = self._shell.run_command(
-            command, timeout=timeout, async_=False
+        # setup-environment will change the current directory to the BitBake
+        # build dir when it is run, so if the user wants to run a command from
+        # somewhere else then we need to handle that after setup-environment
+        # runs. We can't do that with subprocess.run()'s "cwd" option so add a
+        # "cd" command after sourcing setup-environment.
+        cd_command = ""
+        if kwargs.get("cwd"):
+            cd_command = "cd {} &&".format(quote(str(kwargs["cwd"])))
+
+        full_command = "{} && {} {}".format(
+            self._generate_setup_env_command(), cd_command, command
         )
-        if stdout:
-            print(output)
-        print("Done!")
-        return output
-
-    def run_commands(self, commands, timeout=None, stdout=True):
-        """
-        Run a list of commands in the Bitbake environment.
 
-        Mandatory args:
-        * commands (list): the commands to run in the environment
-
-        Optional args:
-        * timeout (int): how long to wait for the next prompt. None means to
-                         wait indefinitely
-        * stdout (bool): flag to print the output to stdout
-
-        If a command fails its execution, it will continue executing the
-        remaining commands.
-
-        """
-        for command in commands:
-            self.run_command(command, timeout=timeout, stdout=stdout)
-
-    def _build_env_variables_string(self):
-        env_string = ""
-        if self.env_variables is not None:
-            env_string = " ".join(
-                "{}={}".format(key, value)
-                for key, value in self.env_variables.items()
-            )
-        return env_string
+        print('Running "{}"...'.format(full_command))
+
+        # Flush stdout and stderr before calling the subprocess so that the
+        # subprocess's output can't appear before prints we've already done.
+        # This is more noticable when stdout and/or stderr are redirected to
+        # files - they are block-buffered rather than line-buffered in that
+        # case (yes, even stderr on at least some versions of Python!).
+        sys.stdout.flush()
+        sys.stderr.flush()
+
+        # Don't modify the caller's dict
+        kwargs = kwargs.copy()
+
+        kwargs["shell"] = False
+        kwargs["cwd"] = str(self.builddir)
+        if "env" in kwargs:
+            kwargs["env"] = kwargs["env"].copy()
```

Comment:
Not sure I get why we copy this again when we copied the entire `dict` a few lines above.

Reply:
My thought process was: the copies are shallow so after the first copy the local `kwargs` `is not` the caller's `kwargs`, but maybe the local `kwargs["env"]` `is` the caller's `kwargs["env"]`. If I want to avoid modifying the caller's `kwargs["env"]` then I need to take a copy of it.